[dynamo][guards] 1/N Guard selectively for DTensor

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #165824

A few internal jobs are observing very high guard overhead for DTensor.
Since we own DTensor, we can make those guards way faster.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

anijain2305 committed:
[dynamo][guards] 1/N Guard selectively for DTensor

A few internal jobs are observing very high guard overhead for DTensor.
Since we own DTensor, we can make those guards way faster.

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guards] 1/N Guard selectively for DTensor"


A few internal jobs are observing very high guard overhead for DTensor.
Since we own DTensor, we can make those guards way faster.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guards] 1/N Guard selectively for DTensor"


A few internal jobs are observing very high guard overhead for DTensor.
Since we own DTensor, we can make those guards way faster.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guards] 1/N Guard selectively for DTensor"


A few internal jobs are observing very high guard overhead for DTensor.
Since we own DTensor, we can make those guards way faster.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guards] 1/N Guard selectively for DTensor"


A few internal jobs are observing very high guard overhead for DTensor.
Since we own DTensor, we can make those guards way faster.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

Lucaskabela reviewed (approved):
Is there a way to easily add test coverage here? Thanks!

anijain2305 committed:
Update on "[dynamo][guards] 1/N Guard selectively for DTensor"


A few internal jobs are observing very high guard overhead for DTensor.
Since we own DTensor, we can make those guards way faster.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

anijain2305 committed:
Update on "[dynamo][guards] 1/N Guard selectively for DTensor"


A few internal jobs are observing very high guard overhead for DTensor.
Since we own DTensor, we can make those guards way faster.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

bdhirsh commented on torch/_dynamo/variables/builder.py:
```diff
@@ -2220,25 +2220,70 @@ def wrap_tensor(self, value: torch.Tensor):
+                if not (
+                    len(flattening_ctx) == 2
+                    and flattening_ctx[0] == value._spec
+                    and flattening_ctx[1] == value.requires_grad
```
it might be a good idea to add a dedicated test that we properly recompile when the requires-gradness of a DTensor input tensor changes

bdhirsh reviewed (approved):
sg!

anijain2305 committed:
Update on "[dynamo][guards] 1/N Guard selectively for DTensor"


A few internal jobs are observing very high guard overhead for DTensor.
Since we own DTensor, we can make those guards way faster.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]

anijain2305 commented:
@pytorchbot revert -m "internal job failed" -c nosignal

anijain2305 commented:
@pytorchbot rebase -b origin/main

anijain2305 committed:
Update on "[dynamo][guards] 1/N Guard selectively for DTensor"


A few internal jobs are observing very high guard overhead for DTensor.
Since we own DTensor, we can make those guards way faster.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames Lucaskabela

[ghstack-poisoned]
